File: src/main/java/me/itzisonn_/meazy/parser/operator/custom/NegationOperator.java
package me.itzisonn_.meazy.parser.operator.custom;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.instruction.NumberType;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression;
import me.itzisonn_.meazy.parser.operator.Operator;
import me.itzisonn_.meazy.parser.operator.OperatorType;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class NegationOperator extends Operator {
    public NegationOperator() {
        super("negation", "-", OperatorType.PREFIX);
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, OperatorExpression operatorExpression) {
        Expression left = operatorExpression.getLeft();
        DataType leftType = left.getType(environment, operatorExpression);

        NumberType leftNumberType = NumberType.valueOf(leftType.getClassDesc());
        if (leftNumberType == null) throw new RuntimeException("Can't negate non-number value");
        if (leftType.isNullable()) throw new RuntimeException("Can't negate nullable number");

        left.emit(instructionsSet, environment, operatorExpression);
        instructionsSet.convertToNumberType(leftNumberType, leftNumberType.unbox());
        instructionsSet.negateNumber(leftType.getClassDesc());
    }

    @Override
    public DataType getType(Environment environment, OperatorExpression operatorExpression) {
        return operatorExpression.getLeft().getType(environment, operatorExpression);
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/runtime/data/operator/custom/NegationOperator.kt
package me.itzisonn_.meazy.runtime.data.operator.custom

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.instruction.NumberType.Companion.valueOf
import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.runtime.data.DataType
import me.itzisonn_.meazy.runtime.data.DataType.Companion.ofNonNull
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression
import me.itzisonn_.meazy.runtime.data.operator.Operator
import me.itzisonn_.meazy.runtime.data.operator.OperatorType
import me.itzisonn_.meazy.runtime.environment.Environment

class NegationOperator : Operator("negation", "-", OperatorType.PREFIX) {
    context(parents: ParentMap)
    override fun emit(instructions: InstructionsSet, environment: Environment, operatorExpression: OperatorExpression) {
        val left = operatorExpression.left
        val leftType = left.getType(environment)

        val leftNumberType = valueOf(leftType.classDesc) ?: error("Can't negate non-number value")
        if (leftType.isNullable) error("Can't negate nullable number")

        left.emit(instructions, environment)
        instructions.convertToNumberType(leftNumberType, leftNumberType.unbox())
        instructions.negateNumber(leftNumberType)
    }

    context(parents: ParentMap)
    override fun getType(environment: Environment, operatorExpression: OperatorExpression): DataType {
        val left = operatorExpression.left
        val leftType = left.getType(environment)

        val leftNumberType = valueOf(leftType.classDesc) ?: error("Can't negate non-number value")
        if (leftType.isNullable) error("Can't negate nullable number")

        return ofNonNull(leftNumberType.unbox().classDesc)
    }
}
